sharktank/sharktank/types/tensors.py

```diff
@@ -274,18 +275,64 @@ def transform_globals(
             prev_globals = next_globals
         return self._clone_with_globals(prev_globals)
 
+    @overload
     def to(
         self,
+        device: str | torch.device | None = None,
+        dtype: torch.dtype | None = None,
+        non_blocking: bool = False,
+        copy: bool = False,
         *,
-        device: Optional[Union[str, torch.device]] = None,
+        memory_format: torch.memory_format = torch.preserve_format,
     ) -> "InferenceTensor":
-        # TODO: reconcile with ops.to(...) and torch.Tensor.to(...).
-        # Do we always want to clone with globals?
-        # This makes our type inconsistent with torch tensors.
-        # If we use this to transform a theta we want to change the theta.
-        # If we want to use this in a computation we don't want to change the theta.
-        return self.transform_globals(
-            lambda d: {k: t.to(device=device) for k, t in d.items()}
+        ...
+
+    @overload
+    def to(
+        self,
+        other: "AnyTensor",
+        non_blocking: bool = False,
+        copy: bool = False,
+        *,
+        memory_format: torch.memory_format = torch.preserve_format,
+    ) -> "InferenceTensor":
+        ...
+
+    @overload
+    def to(
+        self,
+        dtype: torch.dtype,
+        non_blocking: bool = False,
+        copy: bool = False,
+        *,
+        memory_format: torch.memory_format = torch.preserve_format,
+    ) -> "InferenceTensor":
+        ...
+
+    def to(self, *args, **kwargs) -> "InferenceTensor":
+        arg0 = args[0] if len(args) > 0 else None
+        device_overload = ("device" in kwargs) or isinstance(arg0, (str, torch.device))
+        other_overload = ("other" in kwargs) or isinstance(arg0, AnyTensor)
+        memory_overload = ("memory_format" in kwargs) or isinstance(arg0, torch.dtype)
+
+        if device_overload:
+            # Do we always want to clone with globals?
+            # This makes our type inconsistent with torch tensors.
+            # If we use this to transform a theta we want to change the theta.
+            # If we want to use this in a computation we don't want to change the theta.
+            return self.transform_globals(
```

Comment:
The problem here is that `transform_globals` would change to what the external external tensor maps (`ExternalTensorTrait`). Which will be wrong if you use this mid computation in the model and not when doing transformations when preparing a Theta. It will probably manifest rarely as you would need to call this on a Theta tensor that has the external tensor trait.

`transform_globals` would also return a tensor with the same name. The name duplication is maybe fine. In Thetas it is not.

I think we should rename the function when we want to transform thetas and make `to` not be loaded with this additional semantics so it match Torch. This is best to be in another PR.